src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -2513,6 +2515,360 @@ public Expand(Func<TIn, IEnumerator<TOut>> extrapolate)
+                            ca.Invoke(Result.Success(t.Result));
+                        }
+                    };
+
+                private static void CompletionActionVt(object discard)
+                {
+                    var inst = (Holder<T>)discard;
+                    var vtCapture = inst._pending;
+                    inst._pending = default;
+                    if (vtCapture.IsCompletedSuccessfully)
+                    {
+                        inst.Invoke(Result.Success(vtCapture.Result));
+                    }
+                    else if (vtCapture.IsCanceled == false)
+                    {
+                        inst.VTCompletionError(vtCapture);
+                    }
+                }
+
+                private void VTCompletionError(ValueTask<T> vtCapture)
+                {
+                    var t = vtCapture.AsTask();
+                    //We only care about faulted, not canceled.
+                    if (t.IsFaulted)
+                    {
+                        var exception = t.Exception?.InnerExceptions != null &&
+                                        t.Exception.InnerExceptions.Count == 1
+                            ? t.Exception.InnerExceptions[0]
+                            : t.Exception;
+
+                        Invoke(Result.Failure<T>(exception));
+                    }
+                }
+
+                public Holder(Result<T> element, Action<Holder<T>> callback)
+                {
+                    _callback = callback;
+                    Element = element;
+                }
+
+                public void SetElement(Result<T> result)
+                {
+                    Element = result.IsSuccess && result.Value == null
+                        ? Result.Failure<T>(ReactiveStreamsCompliance
+                            .ElementMustNotBeNullException)
+                        : result;
+                }
+
+                public void SetContinuation(ValueTask<T> vt)
+                {
+                    var valueTask = vt;
+                    var peeker =
+                        Unsafe.As<ValueTask<T>, ValueTaskCheatingPeeker<T>>(
+                            ref valueTask);
+                    if (peeker._obj == null)
+                    {
+                        Invoke(Result.Success(peeker._result));
+                    }
+                    else if (peeker._obj is Task<T> asTask)
+                    {
+                        asTask.ContinueWith(TaskCompletedAction, this,
+                            TaskContinuationOptions.NotOnCanceled);
+                    }
+                    else
+                    {
+                        _pending = vt;
+                        var source =
+                            Unsafe.As<IValueTaskSource<T>>(peeker._obj);
+                        source.OnCompleted(OnCompletedAction, this,
+                            peeker._token,
+                            ValueTaskSourceOnCompletedFlags.None);
+                    }
```

Comment:
#### "Magic"

- We get back a `ValueTask<T>`
- Unfortunately, Polite callbacks are not exposed
  - In other words, we can't pass a state in.
- So we just abuse `Unsafe.As` and roll with that.

If this pattern (b/c we do it elsewhere) is overly offensive (I get it) there are some options that come to mind:

1. Always pull the `ValueTask` into a given holder/scope and use a static continuation
2. do a 'pre-check' and branch logic accordingly
  1. I've seen this done in other libs, it's painful for whoever does it the first time but way cleaner than you would think.
  2. Fallback would be some form of compiled expression/etc.
3. Hide all of this behind the newest .NET and rely on `UnsafeAccessor` to get the fields.

Reply:
To use `Unsafe.As`, do we have to set an attribute on the assembly? I think we do this for some `unsafe` stuff inside Akka.Remote already.

Reply:
I don't think so. `Unsafe.As` (and many/most other parts of that API) does not have `unsafe` on the public API method signature.